### PR TITLE
rbd: migration of replication service from controller server

### LINKED ISSUE #3314: rbd: complete migration of volrep to csi-addons
TODO's:

- [x] move logic to `internal/csi-addons/rbd`
- [x] address [comment](https://github.com/ceph/ceph-csi/pull/3608/files#r1092953730)
- [x] remove replication service from controller server
- [x] add minimal unit tests like `internal/csi-addons/rbd/reclaimspace_test.go`
- [x] https://github.com/ceph/ceph-csi/pull/3349

_Originally posted by @nixpanic in https://github.com/ceph/ceph-csi/issues/3305#issuecomment-1216245713_

### PR DESCRIPTION
**Pull request by riya-singhal31**
this commit migrates the replication service from controller server in rbd and adds it to csi-addons.

### CODE REVIEW

File: internal/csi-addons/rbd/replication.go [hunk 1]
```diff
@@ -82,6 +82,12 @@ type ReplicationServer struct {
 	*corerbd.ControllerServer
 }
 
+// NewReplicationServer creates a new IdentityServer which handles
+// the Identity Service requests from the CSI-Addons specification.
```
Comment (Rakshith-R) [inline internal/csi-addons/rbd/replication.go]: This is wrong definition.

File: internal/rbd/driver/driver.go [hunk 2]
```diff
@@ -33,12 +33,10 @@ import (
 
 // Driver contains the default identity,node and controller struct.
 type Driver struct {
-	cd *csicommon.CSIDriver
-
+	cd  *csicommon.CSIDriver
```
Comment (Rakshith-R) [inline internal/rbd/driver/driver.go:36]: unnecessary change.

**Review by Rakshith-R — COMMENTED** [review]
pr title still contains WIP.

Can you test it once to make sure nothing is broken ? 
https://github.com/rook/rook/blob/master/Documentation/Storage-Configuration/Ceph-CSI/ceph-csi-drivers.md#csi-addons-controller

File: internal/rbd/driver/driver.go (hunk 2 above)
Comment (nixpanic) [reply internal/rbd/driver/driver.go:36]: The formatting seems to have been incorrect, this is how it should be, I think?

**Review by nixpanic — COMMENTED** [review]
looks good to me, except for the comment above the `NewReplicationServer()` function.

File: internal/csi-addons/rbd/replication.go (hunk 1 above)
Comment (riya-singhal31) [reply internal/csi-addons/rbd/replication.go]: Yeah, updated. Thanks.

**Comment by riya-singhal31** [comment]
> pr title still contains WIP.
> 
> Can you test it once to make sure nothing is broken ? https://github.com/rook/rook/blob/master/Documentation/Storage-Configuration/Ceph-CSI/ceph-csi-drivers.md#csi-addons-controller

Yes, testing it.

File: internal/csi-addons/rbd/replication.go [hunk 3]
```diff
@@ -82,6 +82,12 @@ type ReplicationServer struct {
 	*corerbd.ControllerServer
 }
 
+// NewReplicationServer creates a new ReplicationServer which handles
+// the Replication Service requests from the CSI-Addons specification.
+func NewReplicationServer() *ReplicationServer {
+	return &ReplicationServer{}
+}
```
Comment (Madhu-1) [inline internal/csi-addons/rbd/replication.go]:
```suggestion
// NewReplicationServer creates a new ReplicationServer which handles
// the Replication Service requests from the CSI-Addons specification.
func NewReplicationServer(c *corerbd.ControllerServer) *ReplicationServer {
	return &ReplicationServer{ControllerServer: c}
}
```

Without the above change, we cannot use the internal locking mechanism. please check this change once again.

**Review by nixpanic — DISMISSED** [review]

File: internal/csi-addons/rbd/replication.go (hunk 3 above)
Comment (Rakshith-R) [reply internal/csi-addons/rbd/replication.go]: Thanks for pointing out madhu ! 

volume locks are initialised way later in New[Controller|Node]Server() calls.

We'll need to move this call 

https://github.com/ceph/ceph-csi/blob/306bfa2a35cda92c9a67b5cbbd879e5c32d8708b/internal/rbd/driver/driver.go#L108-L112

to just before this call to make sure everything is initialized.

https://github.com/ceph/ceph-csi/blob/306bfa2a35cda92c9a67b5cbbd879e5c32d8708b/internal/rbd/driver/driver.go#L182-L183

@riya-singhal31 Please make this change too.

Comment (riya-singhal31) [reply internal/csi-addons/rbd/replication.go]: Updated, thanks!

**Review by nixpanic — APPROVED** [review]

**Comment by Madhu-1** [comment]
@riya-singhal31 Hope this PR is already tested and working as expected?

**Comment by riya-singhal31** [comment]
> @riya-singhal31 Hope this PR is already tested and working as expected?

I'm on it.

**Comment by riya-singhal31** [comment]
Tested. 
Requesting the reviews.

**Comment by riya-singhal31** [comment]
csiaddons server is running as expected and receiving the requests:

1 replication.go:117] ID: 27 Req-ID: 0001-0011-openshift-storage-0000000000000001-e69eee00-3d47-46b5-9e1f-86f3361b16ac mirroringMode is not set in parameters, setting to mirroringMode to default (snapshot)
E0622 16:28:42.441503       1 replication.go:269] ID: 27 Req-ID: 0001-0011-openshift-storage-0000000000000001-e69eee00-3d47-46b5-9e1f-86f3361b16ac failed to enable mirroring on "ocs-storagecluster-cephblockpool/csi-vol-e69eee00-3d47-46b5-9e1f-86f3361b16ac" with error: rbd: ret=-22, Invalid argument
E0622 16:28:42.441559       1 utils.go:210] ID: 27 Req-ID: 0001-0011-openshift-storage-0000000000000001-e69eee00-3d47-46b5-9e1f-86f3361b16ac GRPC error: rpc error: code = Internal desc = failed to enable mirroring on "ocs-storagecluster-cephblockpool/csi-vol-e69eee00-3d47-46b5-9e1f-86f3361b16ac" with error: rbd: ret=-22, Invalid argument
I0622 16:29:02.942045       1 utils.go:195] ID: 28 Req-ID: 0001-0011-openshift-storage-0000000000000001-e69eee00-3d47-46b5-9e1f-86f3361b16ac GRPC call: /replication.Controller/EnableVolumeReplication
I0622 16:29:02.942155       1 utils.go:204] ID: 28 Req-ID: 0001-0011-openshift-storage-0000000000000001-e69eee00-3d47-46b5-9e1f-86f3361b16ac GRPC request: {"parameters":{"schedulingInterval":"1m"},"replication_id":"replicationHandle","secrets":"***stripped***","volume_id":"0001-0011-openshift-storage-0000000000000001-e69eee00-3d47-46b5-9e1f-86f3361b16ac"}
I0622 16:29:02.946989       1 omap.go:88] ID: 28 Req-ID: 0001-0011-openshift-storage-0000000000000001-e69eee00-3d47-46b5-9e1f-86f3361b16ac got omap values: (pool="ocs-storagecluster-cephblockpool", namespace="", name="csi.volume.e69eee00-3d47-46b5-9e1f-86f3361b16ac"): map[csi.imageid:3ad28208e123 csi.imagename:csi-vol-e69eee00-3d47-46b5-9e1f-86f3361b16ac csi.volname:pvc-5bc286d7-19c1-4e24-9ca6-508183d6d585 csi.volume.owner:default]
W0622 16:29:02.972414       1 replication.go:117] ID: 28 Req-ID: 0001-0011-openshift-storage-0000000000000001-e69eee00-3d47-46b5-9e1f-86f3361b16ac mirroringMode is not set in parameters, setting to mirroringMode to default (snapshot)
E0622 16:29:03.020184       1 replication.go:269] ID: 28 Req-ID: 0001-0011-openshift-storage-0000000000000001-e69eee00-3d47-46b5-9e1f-86f3361b16ac failed to enable mirroring on "ocs-storagecluster-cephblockpool/csi-vol-e69eee00-3d47-46b5-9e1f-86f3361b16ac" with error: rbd: ret=-22, Invalid argument
E0622 16:29:03.020236       1 utils.go:210] ID: 28 Req-ID: 0001-0011-openshift-storage-0000000000000001-e69eee00-3d47-46b5-9e1f-86f3361b16ac GRPC error: rpc error: code = Internal desc = failed to enable mirroring on "ocs-storagecluster-cephblockpool/csi-vol-e69eee00-3d47-46b5-9e1f-86f3361b16ac" with error: rbd: ret=-22, Invalid argument

**Review by Rakshith-R — APPROVED** [review]

**Comment by Rakshith-R** [comment]
@mergifyio queue
